bindings/python/src/asyncio.rs

```diff
@@ -0,0 +1,183 @@
+// Copyright 2022 Datafuse Labs
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::collections::HashMap;
+use std::io::SeekFrom;
+use std::str::FromStr;
+use std::sync::Arc;
+
+use ::opendal as od;
+use pyo3::exceptions::PyIOError;
+use pyo3::exceptions::PyNotImplementedError;
+use pyo3::exceptions::PyValueError;
+use pyo3::prelude::*;
+use pyo3::types::PyBytes;
+use pyo3::types::PyDict;
+use pyo3_asyncio::tokio::future_into_py;
+use tokio::io::AsyncReadExt;
+use tokio::io::AsyncSeekExt;
+use tokio::sync::Mutex;
+
+use crate::{build_operator, format_pyerr, Metadata};
+
+#[pyclass(module = "opendal")]
+pub struct AsyncOperator(od::Operator);
+
+#[pymethods]
+impl AsyncOperator {
+    #[new]
+    #[pyo3(signature = (scheme, **map))]
+    pub fn new(scheme: &str, map: Option<&PyDict>) -> PyResult<Self> {
+        let scheme = od::Scheme::from_str(scheme)
+            .map_err(|err| {
+                od::Error::new(od::ErrorKind::Unexpected, "unsupported scheme").set_source(err)
+            })
+            .map_err(format_pyerr)?;
+        let map = map
+            .map(|v| {
+                v.extract::<HashMap<String, String>>()
+                    .expect("must be valid hashmap")
+            })
+            .unwrap_or_default();
+
+        Ok(AsyncOperator(build_operator(scheme, map)?))
+    }
+
+    pub fn read<'p>(&'p self, py: Python<'p>, path: String) -> PyResult<&'p PyAny> {
+        let this = self.0.clone();
+        future_into_py(py, async move {
+            let res: Vec<u8> = this.read(&path).await.map_err(format_pyerr)?;
+            let pybytes: PyObject = Python::with_gil(|py| PyBytes::new(py, &res).into());
+            Ok(pybytes)
+        })
+    }
+
+    pub fn open_reader<'p>(&'p self, py: Python<'p>, path: String) -> PyResult<&'p PyAny> {
+        let this = self.0.clone();
+        future_into_py(py, async move {
+            let reader = this.reader(&path).await.map_err(format_pyerr)?;
```

Comment:
BTW, IMHO the `reader` method shouldn't require `async` at all, it's just creating a `Reader` object.

Reply:
For example, when using fs, the open operation is a system call that must be awaited. And OpenDAL will do something internal jobs at `open` (aka, the `reader` function in OpenDAL).

It's make sense to allow users to use `async with op.open_reader("memory"):`, maybe we can create the real reader until users call `read` or `seek`?

Reply:
The lazy creation sounds great.